flask_appbuilder/security/sqla/apis/user/user.py

```diff
@@ -0,0 +1,200 @@
+from datetime import datetime
+
+from flask import current_app
+from flask import g, request
+from flask_appbuilder import ModelRestApi
+from flask_appbuilder.api import expose, safe
+from flask_appbuilder.const import API_RESULT_RES_KEY
+from flask_appbuilder.models.sqla.interface import SQLAInterface
+from flask_appbuilder.security.decorators import permission_name, protect
+from flask_appbuilder.security.sqla.models import Role, User
+from marshmallow import ValidationError
+from sqlalchemy.exc import IntegrityError
+from werkzeug.security import generate_password_hash
+
+from .schema import UserPostSchema, UserPutSchema
+
+
+class UserApi(ModelRestApi):
+    resource_name = "user"
+    class_permission_name = "User"
+    datamodel = SQLAInterface(User)
+    allow_browser_login = True
+
+    list_columns = [
+        "id",
+        "roles",
+        "first_name",
+        "last_name",
+        "username",
+        "active",
+        "email",
+        "last_login",
+        "login_count",
+        "fail_login_count",
+        "created_on",
+        "changed_on",
+        "created_by.id",
+        "changed_by.id",
+        "created.id",
+        "changed.id",
+    ]
+    show_columns = list_columns
+    add_columns = [
+        "roles",
+        "first_name",
+        "last_name",
+        "username",
+        "active",
+        "email",
+        "password",
+    ]
+    edit_columns = add_columns
+    search_columns = list_columns
+
+    add_model_schema = UserPostSchema()
+    edit_model_schema = UserPutSchema()
+
+    def pre_update(self, item):
+        item.changed_on = datetime.now()
+        item.changed_by_fk = g.user.id
+        if item.password:
+            item.password = generate_password_hash(item.password)
+
+    def pre_add(self, item):
+        item.password = generate_password_hash(item.password)
+
+    @expose("/", methods=["POST"])
+    def post(self):
+        """Create new user
+        ---
+        post:
+          requestBody:
+            description: Model schema
+            required: true
+            content:
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/{{self.__class__.__name__}}.post'
+          responses:
+            201:
+              description: Item changed
+              content:
+                application/json:
+                  schema:
+                    type: object
+                    properties:
+                      result:
+                        $ref: '#/components/schemas/{{self.__class__.__name__}}.post'
+            400:
+              $ref: '#/components/responses/400'
+            401:
+              $ref: '#/components/responses/401'
+            404:
+              $ref: '#/components/responses/404'
+            422:
+              $ref: '#/components/responses/422'
+            500:
+              $ref: '#/components/responses/500'
+        """
+        try:
+            item = self.add_model_schema.load(request.json)
```

Comment:
check https://flask-appbuilder.readthedocs.io/en/latest/rest_api.html#validation-and-custom-validation. Try BaseModelSchema instead of marshmallow schemas, it will give you the SQLAlchemy model for free.

Reply:
but that won't work as we need to populate roles right?

Reply:
but we need to populate roles list first before assiging to model right?